Create new product category to support enrollment codes

vkaracic opened the pull request:
Created a new Client model to use for the enrollment codes. Added a new product class type via a migration. Added a `courses` field to the catalog model because a catalog will have one or more courses and a course will have one or more catalogs. Accompanying tests are included. 

https://openedx.atlassian.net/browse/SOL-1326

@mjfrey @clintonb please take a look.

clintonb commented on ecommerce/extensions/catalogue/models.py:
```diff
@@ -20,6 +20,7 @@ class Catalog(models.Model):
     name = models.CharField(max_length=255)
     partner = models.ForeignKey('partner.Partner', related_name='catalogs')
     stock_records = models.ManyToManyField('partner.StockRecord', blank=True, related_name='catalogs')
+    courses = models.ManyToManyField('courses.Course', blank=True)
```
Why is this necessary? You can get to courses through the stock records and related products. A catalog is the general concept of a catalog. It includes more than just courses, so should share no direct links to courses.

clintonb commented on ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py:
```diff
@@ -0,0 +1,68 @@
+        ('catalogue', '0011_auto_20151019_0639')
+    ]
+    operations = [
+        migrations.RunPython(create_product_class)
```
Include a reverse operation.

mjfrey replied:
Do we need reverse for all migrations?  I do not see one for the Seat migration.

clintonb replied:
Yes, all migrations should have reverse migrations. Feel free to add one for the seat migration. That was created before I was on the team.

clintonb commented on ecommerce/core/models.py:
```diff
@@ -59,3 +59,8 @@ class Meta(object):
+
+
+class Client(User):
+    """Custom user model for clients."""
```
This could use a better docstring. Explain why the model exists.

vkaracic commented:
@clintonb I have reverted the commit that added a new field to Catalog, addressed your other comments and changed the tests accordingly. Can you please take a look now?

mjfrey commented:
LGTM now.
:+1:

clintonb commented on .travis.yml:
```diff
@@ -36,6 +36,6 @@ script:
 branches:
     only:
-        - master
+        - ecodes/master
```
Remove.

clintonb commented on ecommerce/core/models.py:
```diff
@@ -59,3 +59,8 @@ class Meta(object):
+
+
+class Client(User):
+    """The model for the customer that purchases enrollment codes."""
```
`Business clients/purchasers.`

This model doesn't care about enrollment codes, so there should be no mention of enrollment codes in its comment.

clintonb commented on ecommerce/extensions/catalogue/migrations/0002_auto_20150223_1052.py:
```diff
@@ -60,6 +60,19 @@ def create_catalog(apps, schema_editor):
+    ProductClass = apps.get_model("catalogue", "ProductClass")
+
+    Category.objects.filter(slug='seats').delete()
+    ProductAttribute.objects.filter(code='certificate_type').delete()
```
Delete the `ProductClass` and the attributes will be deleted automatically.

clintonb commented on ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py:
```diff
@@ -0,0 +1,81 @@
+    ProductClass = apps.get_model("catalogue", "ProductClass")
+
+    Category.objects.filter(slug='enrollment_code').delete()
+    ProductAttribute.objects.filter(code='end_date').delete()
```
Same as above. You only need to delete the `ProductClass`.

vkaracic commented:
@clintonb updated!

clintonb commented on ecommerce/extensions/catalogue/tests/test_enrollment_codes.py:
```diff
@@ -0,0 +1,39 @@
+        catalog = Catalog.objects.create(
+            partner_id=self.partner.id
+        )
+        course = Course.objects.create(
```
Consider using `CourseFactory`.

clintonb commented on ecommerce/extensions/catalogue/tests/test_enrollment_codes.py:
```diff
@@ -0,0 +1,39 @@
+        )
+
+        ec_product_class = ProductClass.objects.get(slug='enrollment_code')
+        enrollment_code_product = Product.objects.create(
```
Consider using `ProductFactory`.

clintonb commented on ecommerce/extensions/catalogue/tests/test_enrollment_codes.py:
```diff
@@ -0,0 +1,39 @@
+import datetime
+from oscar.core.loading import get_model
```
Imports should be grouped: main lib, third-party, local. Each group should be alphabetized.

clintonb commented:
:+1:

mjfrey commented on ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py:
```diff
@@ -0,0 +1,78 @@
+        image="",
+        name="Enrollment Codes"
+    )
+
```
I am just thinking we missed an attribute we need at fulfillment time.  The type of Voucher to create. Multi-use, Single-use etc...
@clintonb what do you think?
I suggest adding an attribute called "Type" and having it match the ENUM that Abstract Voucher uses.

clintonb replied:
Agreed.

-- 
Clinton Blackburn

vkaracic commented:
@mjfrey @clintonb I have assigned a new product attribute of type option to the enrollment code product class. Please take a look at the last commit. If you are happy with the code I believe I can merge this.

clintonb commented:
:+1:

mjfrey commented:
Great.

:+1:
